code/modules/economy/persistent_bank.dm

```diff
@@ -8,75 +8,109 @@
-
-			//Convert the HELD ITEM string into a purchased item on Mind if applicable here
-			if (p.name == usr.client.persistent_bank_item)
-				var/mob/new_player/playermob = usr.client.mob
+			if(!p.hasJobXP(user.client.key)) continue
+			purchasables += list(
+				list(
+					"pname" = p.name,
+					"cost" = p.cost,
+					"img" = icon2base64(icon(initial(p.icon), initial(p.icon_state), dir=p.icon_dir, frame=p.icon_frame, moving=0)),
+				)
+			)
+			//If you have a purchased item: set held
+			// or you have a persistent item: set it as purchased, and set held
+			var/datum/bank_purchaseable/already_purchased = user.client?.mob?.mind?.purchased_bank_item
+			if (istype(already_purchased) && p.name == already_purchased.name)
+				found_held = p.name
+			else if (p.name == user.client.persistent_bank_item)
+				found_held = p.name
+				var/mob/new_player/playermob = user.client.mob
 				if (playermob.mind)
 					playermob.mind.purchased_bank_item = p
 
-		ret += "<br>This menu is a WIP! Expect additional items and other changes to come in the future!<br/>"
-		return ret
+		if(user.client.persistent_bank_item && user.client.persistent_bank_item != "none" && !found_held)
+			user.client.set_last_purchase(null)
+			boutput(user, "<span class='notice'><b>The thing you previously purchased has been removed from your inventory due to it no longer existing.</b></span>")
+
+		var/truebalance = user.client.persistent_bank
+		if(istype(src.bought_this_round))
+			truebalance += src.bought_this_round.cost
+		. = list(
+			"purchasables" = purchasables,
+			"held" = found_held,
+			"balance" = user.client.persistent_bank,
+			"truebalance" = truebalance
+		)
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		. = ..()
+		if (.)
+			return
+		if(current_state >= GAME_STATE_PLAYING)
+			boutput(ui.user, "<span class='notice'><b>The round has started, you'll have to wait until the next round!</b></span>" )
+			ui.close()
+			return
+		var/id = params["pname"]
+		var/datum/bank_purchaseable/purchased = null
+		for(var/datum/bank_purchaseable/p in persistent_bank_purchaseables)
+			if(id == p.name)
+				purchased = p
 
-	OnClick( var/client/who, var/id )
-		if (istext(id))
-			id = text2num(id)
-		var/datum/bank_purchaseable/p = persistent_bank_purchaseables[id]
-		if( p )
-			if (try_purchase(who, p))
-				Unsubscribe( who )
+		if(istype(purchased))
+			if (try_purchase(ui.user.client, purchased))
+				ui.close()
 		else
-			boutput( who, "<span class='notice'><b>Oh no! Something is broken. Please tell a coder. (problem retrieving purchaseable id : [id])</b></span>" )
+			boutput( ui.user, "<span class='notice'><b>Oh no! Something is broken. Please tell a coder. (problem retrieving purchaseable id : [id])</b></span>" )
 
 	proc/try_purchase(var/client/c, var/datum/bank_purchaseable/p)
+		if(current_state >= GAME_STATE_PLAYING) //if the game has started, you can't buy things
+			return FALSE
```

Comment:
Possibly bad logic; I think people joining mid-round can still make use of Spacebux purchases?